docker/build/installers/py3_requirements.txt

```diff
@@ -19,6 +19,10 @@ simplejson
 pyproj
 shapely
 matplotlib
+python_dev_tools
```

Comment:
It seems that python_dev_tools is a big module that have so many dependencies. Does we really require them all ?

Reply:
> It seems that python_dev_tools is a big module that have so many dependencies. Does we really require them all ?

```
Successfully installed GitPython-3.1.7 Jinja2-2.11.2 Pygments-2.6.1 SecretStorage-3.1.2 Sphinx-2.3.1 alabaster-0.7.12 appdirs-1.4.4 attrs-19.3.0 autoflake-1.3.1 babel-2.8.0 bandit-1.6.2 black-19.10b0 bleach-3.1.5 bumpversion-0.5.3 click-7.1.2 coverage-5.0.3 coveralls-1.10.0 distlib-0.3.1 doc8-0.8.0 docopt-0.6.2 docutils-0.16 entrypoints-0.3 filelock-3.0.12 flake8-3.7.9 flake8-2020-1.5.0 flake8-bandit-2.1.2 flake8-broken-line-0.1.1 flake8-bugbear-20.1.4 flake8-builtins-1.4.2 flake8-comprehensions-3.2.2 flake8-debugger-3.2.1 flake8-fixme-1.1.1 flake8-isort-2.8.0 flake8-logging-format-0.6.0 flake8-mutable-1.2.0 flake8-polyfill-1.0.2 flake8-variables-names-0.0.3 gitdb-4.0.5 imagesize-1.2.0 importlib-metadata-1.7.0 importlib-resources-3.0.0 isort-4.3.21 jeepney-0.4.3 keyring-21.2.1 more-itertools-8.4.0 packaging-20.4 pathspec-0.8.0 pbr-5.4.5 pep8-naming-0.9.1 pip-20.0.2 pkginfo-1.5.0.1 pluggy-0.13.1 py-1.9.0 pycodestyle-2.5.0 pydocstyle-5.0.2 pyflakes-2.1.1 pyparsing-2.4.7 pytest-5.3.5 pytest-cov-2.8.1 pytest-runner-5.2 python-dev-tools-2020.2.5 pyupgrade-1.26.2 readme-renderer-26.0 regex-2020.7.14 requests-toolbelt-0.9.1 restructuredtext-lint-1.3.1 smmap-3.0.4 snowballstemmer-2.0.0 sphinxcontrib-applehelp-1.0.2 sphinxcontrib-devhelp-1.0.2 sphinxcontrib-htmlhelp-1.0.3 sphinxcontrib-jsmath-1.0.1 sphinxcontrib-qthelp-1.0.3 sphinxcontrib-serializinghtml-1.1.4 stevedore-3.1.0 testfixtures-6.14.1 tokenize-rt-4.0.0 toml-0.10.1 tox-3.14.3 tqdm-4.48.0 twine-3.1.1 typed-ast-1.4.1 virtualenv-20.0.27 wcwidth-0.2.5 webencodings-0.5.1 wheel-0.34.2 zipp-3.1.0
```

Reply:
Agree! Prefer exact list of individual libs.

Reply:
Well it's the dependency of another module `pyaudio`. I will check it later

Reply:
I find another source of installation -- `sudo apt-get install python3-pyaudio`, which will cost (4961 Kb for all)
```shell
The following NEW packages will be installed:
  libasound2 libasound2-data libjack-jackd2-0 libportaudio2 libsamplerate0 python3-pyaudio
0 upgraded, 6 newly installed, 0 to remove and 41 not upgraded.
Need to get 1688 kB of archives.
After this operation, 4961 kB of additional disk space will be used.
```
Compared to install using pip, which requires
`sudo apt-get install -y portaudio19-dev` && `pip install python_dev_tools, pyaudio`
I test `python_dev_tools` on a clean virtual python environment and it cost about 100M (though it maybe because it's too clean that some common dependencies are counted too)
![image](https://user-images.githubusercontent.com/20517842/87809383-4f6d5100-c829-11ea-8668-55d48a2874ae.png)
Does it sound good to you to change it to apt-get?

Reply:
> I find another source of installation -- `sudo apt-get install python3-pyaudio`, which will cost (4961 Kb for all)
> 
> ```shell
> The following NEW packages will be installed:
>   libasound2 libasound2-data libjack-jackd2-0 libportaudio2 libsamplerate0 python3-pyaudio
> 0 upgraded, 6 newly installed, 0 to remove and 41 not upgraded.
> Need to get 1688 kB of archives.
> After this operation, 4961 kB of additional disk space will be used.
> ```
> Does it sound good to you to change it to apt-get?

Yep. As long as it works for the relevant module (audio driver in your case).